Feature - Make Promoted Search Results Orderable

lb- opened the pull request:
Extends the work done on #4614 
Closes #3897

Thanks @chrisranjana - really awesome stuff.

Just adds some tests, a bit of formatting in the Django template files and a bit of a reorganisation of how the ordering url args are handled in the view.

* Do the tests still pass? ✅ 
* Does the code comply with the style guide?  ✅ 
* For Python changes: Have you added tests to cover the new/fixed behaviour?  ✅ 
* For front-end changes: N/A
* For new features: Has the documentation been updated accordingly? N/A - pretty much expected behaviour and other similar pages are not documented in this way.

chosak commented on wagtail/contrib/search_promotions/views.py:
```diff
@@ -19,16 +20,30 @@
-    is_searching = False
-    query_string = request.GET.get('q', "")
+    # Ordering
+    valid_ordering = ['query_string', '-query_string', 'ohits', '-ohits']
```
At first read the meaning of the query string parameter "ohits" isn't really clear. Since under the hood this is an annotated value that doesn't have to line up to an existing model field, it might be nicer if the query string parameter matched what is displayed in the column header, maybe "views".

lb- replied:
@chosak  - thanks for the feedback, I agree that ohits was a bit confusing. I have renamed to `views`

chosak commented on wagtail/contrib/search_promotions/views.py:
```diff
@@ -19,16 +20,30 @@
+    queries = queries.order_by(ordering)
 
     # Search
+    is_searching = False
```
The search behavior doesn't work quite as expected for me, because the AJAX call loses the ordering once you start typing.

The search behavior is a little inconsistent, although this appears to be common in other parts of Wagtail (for example the user search). I didn't see this behavior mentioned on #5191.

If you search first and then order, you lose the search term:

1. Search first, URL is http://localhost:8000/admin/searchpicks/?q=b.
2. Click on ordering, link goes to http://localhost:8000/admin/searchpicks/?ordering=ohits, losing the search.

If you order first and then search, you lose the order:

1. Order first, URL is http://localhost:8000/admin/searchpicks/?ordering=ohits.
2. Start typing a search, AJAX request is made to http://localhost:8000/admin/searchpicks/?q=b (without the ordering param), so you lose the order.

To me the expected behavior is that searches should definitely persist across ordering. I'm not sure about the other way, whether ordering should persist if you do a new search, but I think maybe it should, as the user has indicated a preferred ordering (in this case, a common one would probably be to order by most common search term).

This doesn't need to hold up this PR, but I'm curious what people think is the best way to handle this.

lb- replied:
@chosak  - thanks for that, I was aware of this issue and intentionally have not fixed it in this PR. I assumed, wrongly, an issue had already been created for this (I thought I had created one a year or so ago). I searched but could not find it.

I have now created a new issue that contains your notes and a bit more.

https://github.com/wagtail/wagtail/issues/5529

chosak commented on wagtail/contrib/search_promotions/tests.py:
```diff
@@ -154,6 +154,70 @@ def test_results_are_ordered_alphabetically(self):
+            description="An odly popular search term?",
+        )
+
+        popularQuery = Query.get("sub-optimal")
```
Looks like this needs to be changed to match what the tests expect (or vice-versa):

```suggestion
        popularQuery = Query.get("suboptimal")
```

gasman replied:
Sounds possibly related to #5416 / #5418 - perhaps this test was written before that change was committed?

lb- commented:
I'll take a look - I might need to merge in the latest from master.

lb- commented:
@chosak  - can you take a look at this and approve if you think this is good to go, I am happy to merge this in but just wanted to make sure this gets another check through

lb- commented:
@chosak thanks for the suggestion, I have now added this (although I used `title` as this seemed more appropriate and useful), although this made me start down rabbit hole of accessibility support for `title` vs `aria-label`. It appears that this needs a fair bit of work throughout Wagtail (I know there are other issues for this).

Plus the way we are doing icons, as you probably know, seems to cause some issues.

happy to change back to aria-label if you think that is better, or use both (again there are various potential issues with that).

Note: I have also squashed, rebased and put all into one commit

lb- commented:
@chosak did a bit more digging and testing, revised to use `aria-label` as per your original suggestion.

lb- commented:
Merged in via 1f421e35f00b5463844171047bd099259c11e95b

Thanks again @chrisranjana  for your original work on this
